libbpf-tools/biosnoop.c

```diff
@@ -0,0 +1,295 @@
+
+static struct env {
+	char *disk;
+	int disk_len;
+	int duration;
+	bool timestamp;
+	bool queued;
+	bool verbose;
+} env = {};
+
+static volatile __u64 start_ts;
+
+const char *argp_program_version = "biosnoop 0.1";
+const char *argp_program_bug_address = "<ethercflow@gmail.com>";
+const char argp_program_doc[] =
+"Summarize block device I/O latency as a histogram.\n"
+"\n"
+"USAGE: biosnoop [-h] [-T] [-Q]\n"
+"\n"
+"EXAMPLES:\n"
+"    biosnoop              # summarize block I/O latency as a histogram\n"
+"    biosnoop -Q           # include OS queued time in I/O time\n"
+"    biosnoop 10           # trace for 10 seconds only\n"
+"    biosnoop -d sdc       # trace sdc only\n";
+
+static const struct argp_option opts[] = {
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{ "queued", 'Q', NULL, 0, "Include OS queued time in I/O time" },
+	{ "disk",  'd', "DISK",  0, "Trace this disk only" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'h':
+		argp_usage(state);
+		break;
+	case 'Q':
+		env.queued = true;
+		break;
+	case 'd':
+		env.disk = arg;
+		env.disk_len = strlen(arg) + 1;
+		if (env.disk_len > DISK_NAME_LEN) {
+			fprintf(stderr, "invaild disk name: too long\n");
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.duration = strtoll(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void blk_fill_rwbs(char *rwbs, unsigned int op)
+{
+	int i = 0;
+
+	if (op & REQ_PREFLUSH)
+		rwbs[i++] = 'F';
+
+	switch (op & REQ_OP_MASK) {
+	case REQ_OP_WRITE:
+	case REQ_OP_WRITE_SAME:
+		rwbs[i++] = 'W';
+		break;
+	case REQ_OP_DISCARD:
+		rwbs[i++] = 'D';
+		break;
+	case REQ_OP_SECURE_ERASE:
+		rwbs[i++] = 'D';
+		rwbs[i++] = 'E';
+		break;
+	case REQ_OP_FLUSH:
+		rwbs[i++] = 'F';
+		break;
+	case REQ_OP_READ:
+		rwbs[i++] = 'R';
+		break;
+	default:
+		rwbs[i++] = 'N';
+	}
+
+	if (op & REQ_FUA)
+		rwbs[i++] = 'F';
+	if (op & REQ_RAHEAD)
+		rwbs[i++] = 'A';
+	if (op & REQ_SYNC)
+		rwbs[i++] = 'S';
+	if (op & REQ_META)
+		rwbs[i++] = 'M';
+
+	rwbs[i] = '\0';
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	char rwbs[RWBS_LEN];
+
+	if (!start_ts)
+		start_ts = e->ts;
+	blk_fill_rwbs(rwbs, e->cmd_flags);
+	printf("%-11.6f %-14.14s %-6d %-7s %-4s %-10lld %-7d ",
+		((double)e->ts - start_ts) / 1000000000,
```

Comment:
I might have commented on similar issue in another PR, but I think this looks cleaner (and is potentially a bit more precises, not that it matters in this case, though):

```
(e->ts - start_s) / 1000000000.0
```

Reply:
Oh, I remember that you did give advice in another PR, I forgot to check this PR, I should be more rigorous. Fixed, thank.